README.md
PackyTacky
==========

A simple game made in Lua with the Love2D framework (0.9.1).

PackyTacky is a simple 2D game in which the player controls a square shaped character that has to move around and eat red squares to increase the score. There is also an enemy character which looks similar to the player's character but is different in color. He is attracted towards the food. If the enemy character eats the food, the score decreases. If the enemy eats the player character, the game ends.

Please note that this is a work in progress. This is just my implementation of the idea. There is much room for improvement and additions.

This game is distributed under the Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International Public License. In summary, it means you can copy and redistribute this game or remix, transform or build upon it after giving all due credits. You must not use this game or any of its material for commercial purposes. Upon redistribution of this or a remix or transformed version of this game, you must share it with the same license and give due credit.

As of yet, the game has been released for Windows 32 and 64 bits, and Mac OSX (Credits go to Eamonn Rea https://github.com/sonic2kk)

Download The Game:
----
To play the game, you can download the binaries for Windows 32 bit at http://goo.gl/lqxNU4 and for Windows 64 bit at http://goo.gl/WhSEpL

Download for Mac OSX: http://goo.gl/sjEb8A

Screenshots:
----
![Title](Screenshots/title.png "Title Screen")

![Gameplay](Screenshots/game.png "In Game")

![Game Over](Screenshots/gameover.png "Game Over")


CHANGELOG.md
ChangeLog
==========

v2.0 (August 4, 2014)
---
- FIXED: Two or more apples spawning at the same place one after the other
- Minor optimizations
- Renamed "startmenu.lua" to "menu.lua"
- Code cleanup
- Modified enemy speed increments
- Added option to take screenshot at game over
- Added temporary speed increasing powerups
- Released for Mac OSX

v1.2 (July 7, 2014)
----
- FIXED : Bugged reset highscore button
- FIXED : Enemy AI improved
- Customized buttons
- Added the option to mute the background music in start and pause menu
- Enemy speed increases as score increases to increase difficulty

v1.1 (July 6, 2014)
----
- FIXED : Bugged diagonal movement of player which gave player more speed than moving straight
- Improved Graphics; player sprite, enemy sprite and food sprite
- Improved animations
- Listed libraries and their creators in the credits section
- Option to reset highscore at the start menu

v1.0 (July 5, 2014)
----
- Initial release
- Released for win32 and win64

food.lua
Food = {}

function Food:new()
	f = {w=16, h=16, x=0, y=0, foodImage = gr.newImage("img/apple.png"), powerupImage = gr.newImage("img/pwr.png"), powerUp = false}
	Food:respawn()
	setmetatable(f, {__index = Food})
	return f
end

function Food:update(dt)
	-- checking collision with player
	if o.y+o.h > f.y and o.x+o.w > f.x and o.y < f.y+f.h and o.x < f.x+f.w then
		o.score = o.score + 1
		if o.score - prevScore == 10 and e.speed < 500 then
			prevScore = o.score
			e.speed = e.speed + 50
		end
		
		if (f.powerUp == true and o.score > 1) then
			f.powerUp = false
			o.speedup = true
			timerS = 0 --set timer start to zero
			pwrSfx:play() --if apple is a powerup, play the powerup sound
		else foodSfx:play() --if the apple is a red one, play the regular sound
		end
		Food:respawn()
	end
	-- checking collision with enemy character
	if e.y+e.h > f.y and e.x+e.w > f.x and e.y < f.y+f.h and e.x < f.x+f.w then
		enemySfx:play()
		if o.score > 0 then
			o.score = o.score - 1
		end

		f.powerUp = false
		Food:respawn()
	end
end

function Food:respawn()
	r = math.random(1,15)
	if (r == 3 and f.powerUp == false) then --1/15 chance of every apple being a powerup
		f.powerUp = true
	end
    local lastX, lastY = f.x, f.y
    f.x = math.random(gr.getWidth() - f.w)
    f.y = math.random(32, gr.getHeight() - f.h)
    if f.x==lastX and f.y==lastY then
        f.x = (f.x + math.random(gr.getWidth()-f.w) + 5) % gr.getWidth()
	end
end

function Food:draw()
	gr.setColor(255, 255, 255)
	if (f.powerUp == false or o.score == 0) then --if the apple is not a powerup
		gr.draw(f.foodImage, f.x, f.y) --we draw the regular red apple
	else gr.draw(f.powerupImage, f.x, f.y) --if it is, we draw the green apple
	end
end

enemy.lua
Enemy = {}

local gameoverPlayed = false

function math.angle(x1,y1, x2,y2)
	return math.atan2(y2-y1, x2-x1)
end

function Enemy:new()
	e = {x=gr.getWidth()+30, y=100, w=30, h=30, facing="right", imgLeft=gr.newImage("img/enLeft.png"),
		imgRight=gr.newImage("img/enRight.png"), speed=200, angleDivider=1}
	Enemy:respawn()
	setmetatable(e, {__index = Enemy})
	return e
end

function Enemy:update(dt)
	--checking collision with player
	if o.y+o.h > e.y and o.x+o.w > e.x and o.y < e.y+e.h and o.x < e.x+e.w then
		-- this nested condition checks if the gameover sound has been already played once
		-- if this is not implemented, the gameover sound will be played over and over again
		-- and the game will keep sleeping because of love.timer.sleep()
		if gameoverPlayed == false then
			playingBgm:stop()
			gameoverMusic:play()
			love.timer.sleep(2)
			gameoverPlayed = true
		end
		gamestate = "gameover"
	end

	if gamestate == "playing" and e.facing == "left" and e.x > -100 then
		e.x = e.x - e.speed * dt
		_angle = math.angle(e.x, e.y, f.x, f.y) -- gets the angle between the spawned food and enemy character
		-- adjusts angle toward the spawned food
		dy = math.sin(_angle) * (dt * (e.speed/e.angleDivider)) -- we can use angleDivider to vary difficulty
		-- moves in the direction of the calculated angle
		e.y = e.y + dy
	elseif gamestate == "playing" and e.x <= -100 then
		Enemy:respawn()
		e.facing = "right"
	end
	
	if e.facing == "right" and gamestate == "playing" and e.x < gr.getWidth() + 100 then
		e.x = e.x + e.speed * dt
		-- same calculations as done for e.facing == "left"
		_angle = math.angle(e.x, e.y, f.x, f.y)
		dy = math.sin(_angle) * (dt * (e.speed/e.angleDivider))
		e.y = e.y + dy
	elseif gamestate == "playing" and e.x >= gr.getWidth() + 100 then
		Enemy:respawn()
		e.facing = "left"
	end
end

function Enemy:respawn()
	e.y = math.random(32, gr.getHeight() - 32)
end

function Enemy:newGame()
	e.speed = 200
	gameoverPlayed = false
	e.x = gr.getWidth()+30
	Enemy:respawn()
end

player.lua
Player = {}
local dx, dy = 0, 0

function Player:new()
	o = {x=gr.getWidth()/2, y=gr.getHeight()/2, w=30, h=30, score=0, highscore=0,
		imgUp=gr.newImage("img/up.png"), imgDown=gr.newImage("img/down.png"),
		imgLeft=gr.newImage("img/left.png"), imgRight=gr.newImage("img/right.png"),
		facing="left", speed = 300, speedup = false}
	setmetatable(o, {__index = Player})
	return o
end

function Player:spawnAgain()
	-- Spawning the player in the center of the playing area
	o.x = gr.getWidth()/2
	o.y = gr.getHeight()/2
	o.facing = "left"
	o.speed = 300
	-- Resetting score
	score = 0
end

function Player:update(dt)
	if (o.speedup == true) then --if player eats a speed powerup
		o.speed = 500 --increase speed
		timerS = timerS + dt --update timer
		if (timerS > 5) then --when timer is over 5 seconds
			o.speedup = false --reset everything
			o.speed = 300
			timerS = 0
		end
	end

	dx, dy = 0, 0 --delta x and y

	--MOVEMENT CONTROLS

	if (kb.isDown("a") or kb.isDown("left")) and o.x > 0 and gamestate == "playing" then
		dx = dx - o.speed * dt
		o.facing = "left"
	end
	if kb.isDown("s") or kb.isDown ("down") and gamestate == "playing" then
		dy = dy + o.speed * dt
		o.facing = "down"
	end
	if (kb.isDown("w") or kb.isDown("up")) and o.y > 32 and gamestate == "playing" then
		dy = dy - o.speed * dt
		o.facing = "up"
	end
	if kb.isDown("d") or kb.isDown("right") and gamestate == "playing" then
		dx = dx + o.speed * dt
		o.facing = "right"
	end

	-- To keep velocity constant when player moves diagonally, we use Pythagorean theorem
	-- Pythagorean theorem is hypotenuse^2 = base^2 + height^2
	-- So we need the ratio dx/(sqroot(2)) assuming base = 1 and height = 1, so hypotenuse^2 = 2
	if dx ~= 0 and dy ~= 0 and gamestate == "playing" then
    	dx = dx / 1.41421 -- 1.41421 is an estimation of square root of 2
    	dy = dy / 1.41421
	end
   	o.x = o.x + dx
   	o.y = o.y + dy

   	-- Checking if player collides with the boundaries of the playing area
   	-- stops if colliding
	if o.y > gr.getHeight() - o.h then
		o.y = gr.getHeight() - o.h
	end
	if o.x > gr.getWidth() - o.w then
		o.x = gr.getWidth() - o.w
	end
end

startmenu.lua
function render_menu()
	--MAIN MENU

	local muteButton = loveframes.Create("imagebutton")
	muteButton:SetSize(100, 30)
	muteButton:SetImage("img/muteButton.png")
	muteButton:SetPos(5, 5)
	muteButton:SetState("startmenu")
	muteButton:SetText("")
	muteButton.OnClick = function()
		if mute == false then
			mute = true
			menuMusic:stop()
		else
			mute = false
			menuMusic:play()
		end
	end

	local startButton = loveframes.Create("imagebutton")
	startButton:SetSize(150, 40)
	startButton:SetImage("img/playButton.png")
	startButton:SetPos((gr.getWidth()-150)/2 , (gr.getHeight()-160)/2)
	startButton:SetState("startmenu")
	startButton:SetText("")
	startButton.OnClick = function()
		p.score = 0
		p:spawnAgain()
		f:respawn()
		en:newGame()
		gamestate = "playing"
		loveframes.SetState("playing")
		menuMusic:stop()
	end

	local instructButton = loveframes.Create("imagebutton")
	instructButton:SetSize(150, 40)
	instructButton:SetImage("img/instButton.png")
	instructButton:SetPos((gr.getWidth()-150)/2 , (gr.getHeight()-40)/2)
	instructButton:SetState("startmenu")
	instructButton:SetText("")
	instructButton.OnClick = function()
		gamestate = "help"
		loveframes.SetState("help")
	end

	local creditsButton = loveframes.Create("imagebutton")
	creditsButton:SetSize(150, 40)
	creditsButton:SetImage("img/creditsButton.png")
	creditsButton:SetPos((gr.getWidth()-150)/2 , (gr.getHeight()+80)/2)
	creditsButton:SetState("startmenu")
	creditsButton:SetText("")
	creditsButton.OnClick = function()
		gamestate = "credits"
		loveframes.SetState("credits")
	end

	local quitButton = loveframes.Create("imagebutton")
	quitButton:SetSize(150, 40)
	quitButton:SetImage("img/quitButton.png")
	quitButton:SetPos((gr.getWidth()-150)/2 , (gr.getHeight()+200)/2)
	quitButton:SetState("startmenu")
	quitButton:SetText("")
	quitButton.OnClick = function()
		love.event.quit()
	end

	local resetHSButton = loveframes.Create("imagebutton")
	resetHSButton:SetSize(100, 30)
	resetHSButton:SetImage("img/resetButton.png")
	resetHSButton:SetPos(gr.getWidth()/2 + 35, gr.getHeight()/4 + 15) 
	resetHSButton:SetState("startmenu")
	resetHSButton:SetText("")
	resetHSButton.OnClick = function()
		p.score = 0
		p.highscore = 0
		fs.write("scores.lua", "p.highscore\n=\n0")
	end

	--PAUSE MENU

	local muteButtonP = loveframes.Create("imagebutton")
	muteButtonP:SetSize(100, 30)
	muteButtonP:SetImage("img/muteButton.png")
	muteButtonP:SetPos(5, 5)
	muteButtonP:SetState("paused")
	muteButtonP:SetText("")
	muteButtonP.OnClick = function()
		if mute == false then
			mute = true
		else
			mute = false
		end
	end

	local resumeButton = loveframes.Create("imagebutton")
	resumeButton:SetSize(150, 40)
	resumeButton:SetImage("img/resumeButton.png")
	resumeButton:SetPos((gr.getWidth()-150)/2 , (gr.getHeight()-160)/2)
	resumeButton:SetState("paused")
	resumeButton:SetText("")
	resumeButton.OnClick = function()
		loveframes.SetState("playing")
		gamestate = "playing"
	end

	local backButton = loveframes.Create("imagebutton")
	backButton:SetSize(150, 40)
	backButton:SetImage("img/backButton.png")
	backButton:SetPos((gr.getWidth()-150)/2 , (gr.getHeight()-40)/2)
	backButton:SetState("paused")
	backButton:SetText("")
	backButton.OnClick = function()
		loveframes.SetState("startmenu")
		gamestate = "startmenu"
		playingBgm:stop()
		if mute == false then menuMusic:play() end
	end

	local quitButtonP = loveframes.Create("imagebutton")
	quitButtonP:SetSize(150, 40)
	quitButtonP:SetImage("img/quitPButton.png")
	quitButtonP:SetPos((gr.getWidth()-150)/2 , (gr.getHeight()+80)/2)
	quitButtonP:SetState("paused")
	quitButtonP:SetText("")
	quitButtonP.OnClick = function()
		love.event.quit()
	end

	--CREDITS MENU AND INFO MENU

	local backButtonC = loveframes.Create("imagebutton")
	backButtonC:SetSize(150, 40)
	backButtonC:SetImage("img/backButton.png")
	backButtonC:SetPos((gr.getWidth()-150)/2 , (gr.getHeight()+160)/2)
	backButtonC:SetState("credits")
	backButtonC:SetText("")
	backButtonC.OnClick = function()
		loveframes.SetState("startmenu")
		gamestate = "startmenu"
	end

	local backButtonI = loveframes.Create("imagebutton")
	backButtonI:SetSize(150, 40)
	backButtonI:SetImage("img/backButton.png")
	backButtonI:SetPos((gr.getWidth()-150)/2 , (gr.getHeight()+220)/2)
	backButtonI:SetState("help")
	backButtonI:SetText("")
	backButtonI.OnClick = function()
		loveframes.SetState("startmenu")
		gamestate = "startmenu"
	end

	--GAME OVER MENU

	local backButtonG = loveframes.Create("imagebutton")
	backButtonG:SetSize(150, 40)
	backButtonG:SetImage("img/backButton.png")
	backButtonG:SetPos((gr.getWidth()-150)/2 , (gr.getHeight()+200)/2)
	backButtonG:SetState("gameover")
	backButtonG:SetText("")
	backButtonG.OnClick = function()
		p:spawnAgain()
		en:newGame()
		gamestate = "startmenu"
		loveframes.SetState("startmenu")
		if mute == false then menuMusic:play() end
	end

	local tryAgainButton = loveframes.Create("imagebutton")
	tryAgainButton:SetSize(150, 40)
	tryAgainButton:SetImage("img/tryButton.png")
	tryAgainButton:SetPos((gr.getWidth()-150)/2 , (gr.getHeight()+80)/2)
	tryAgainButton:SetState("gameover")
	tryAgainButton:SetText("")
	tryAgainButton.OnClick = function()
		p:spawnAgain()
		p.score = 0
		f:respawn()
		e:newGame()
		gamestate = "playing"
		loveframes.SetState("playing")
		menuMusic:stop()
	end
end

menu.lua
function saveSS()
	local file = io.open (fs.getSaveDirectory() .. "/score.png", "rb")
	local inp = file:read("*a")
	file:close()

	local outfile = io.open("score.png", "wb")
	outfile:write(inp)
	outfile:close()
end

function render_menu()
	--MAIN MENU

	local muteButton = loveframes.Create("imagebutton")
	muteButton:SetSize(100, 30)
	muteButton:SetImage("img/muteButton.png")
	muteButton:SetPos(5, 5)
	muteButton:SetState("startmenu")
	muteButton:SetText("")
	muteButton.OnClick = function()
		if mute == false then
			mute = true
			menuMusic:stop()
		else
			mute = false
			menuMusic:play()
		end
	end

	local startButton = loveframes.Create("imagebutton")
	startButton:SetSize(150, 40)
	startButton:SetImage("img/playButton.png")
	startButton:SetPos((gr.getWidth()-150)/2 , (gr.getHeight()-160)/2)
	startButton:SetState("startmenu")
	startButton:SetText("")
	startButton.OnClick = function()
		p.score = 0
		p:spawnAgain()
		f:respawn()
		en:newGame()
		gamestate = "playing"
		loveframes.SetState("playing")
		menuMusic:stop()
	end

	local instructButton = loveframes.Create("imagebutton")
	instructButton:SetSize(150, 40)
	instructButton:SetImage("img/instButton.png")
	instructButton:SetPos((gr.getWidth()-150)/2 , (gr.getHeight()-40)/2)
	instructButton:SetState("startmenu")
	instructButton:SetText("")
	instructButton.OnClick = function()
		gamestate = "help"
		loveframes.SetState("help")
	end

	local creditsButton = loveframes.Create("imagebutton")
	creditsButton:SetSize(150, 40)
	creditsButton:SetImage("img/creditsButton.png")
	creditsButton:SetPos((gr.getWidth()-150)/2 , (gr.getHeight()+80)/2)
	creditsButton:SetState("startmenu")
	creditsButton:SetText("")
	creditsButton.OnClick = function()
		gamestate = "credits"
		loveframes.SetState("credits")
	end

	local quitButton = loveframes.Create("imagebutton")
	quitButton:SetSize(150, 40)
	quitButton:SetImage("img/quitButton.png")
	quitButton:SetPos((gr.getWidth()-150)/2 , (gr.getHeight()+200)/2)
	quitButton:SetState("startmenu")
	quitButton:SetText("")
	quitButton.OnClick = function()
		love.event.quit()
	end

	local resetHSButton = loveframes.Create("imagebutton")
	resetHSButton:SetSize(100, 30)
	resetHSButton:SetImage("img/resetButton.png")
	resetHSButton:SetPos(gr.getWidth()/2 + 35, gr.getHeight()/4 + 15) 
	resetHSButton:SetState("startmenu")
	resetHSButton:SetText("")
	resetHSButton.OnClick = function()
		p.score = 0
		p.highscore = 0
		fs.write("scores.lua", "p.highscore\n=\n0")
	end

	--PAUSE MENU

	local muteButtonP = loveframes.Create("imagebutton")
	muteButtonP:SetSize(100, 30)
	muteButtonP:SetImage("img/muteButton.png")
	muteButtonP:SetPos(5, 5)
	muteButtonP:SetState("paused")
	muteButtonP:SetText("")
	muteButtonP.OnClick = function()
		if mute == false then
			mute = true
		else
			mute = false
		end
	end

	local resumeButton = loveframes.Create("imagebutton")
	resumeButton:SetSize(150, 40)
	resumeButton:SetImage("img/resumeButton.png")
	resumeButton:SetPos((gr.getWidth()-150)/2 , (gr.getHeight()-160)/2)
	resumeButton:SetState("paused")
	resumeButton:SetText("")
	resumeButton.OnClick = function()
		loveframes.SetState("playing")
		gamestate = "playing"
	end

	local backButton = loveframes.Create("imagebutton")
	backButton:SetSize(150, 40)
	backButton:SetImage("img/backButton.png")
	backButton:SetPos((gr.getWidth()-150)/2 , (gr.getHeight()-40)/2)
	backButton:SetState("paused")
	backButton:SetText("")
	backButton.OnClick = function()
		loveframes.SetState("startmenu")
		gamestate = "startmenu"
		playingBgm:stop()
		if mute == false then menuMusic:play() end
	end

	local quitButtonP = loveframes.Create("imagebutton")
	quitButtonP:SetSize(150, 40)
	quitButtonP:SetImage("img/quitPButton.png")
	quitButtonP:SetPos((gr.getWidth()-150)/2 , (gr.getHeight()+80)/2)
	quitButtonP:SetState("paused")
	quitButtonP:SetText("")
	quitButtonP.OnClick = function()
		love.event.quit()
	end

	--CREDITS MENU AND INFO MENU

	local backButtonC = loveframes.Create("imagebutton")
	backButtonC:SetSize(150, 40)
	backButtonC:SetImage("img/backButton.png")
	backButtonC:SetPos((gr.getWidth()-150)/2 , (gr.getHeight()+160)/2)
	backButtonC:SetState("credits")
	backButtonC:SetText("")
	backButtonC.OnClick = function()
		loveframes.SetState("startmenu")
		gamestate = "startmenu"
	end

	local backButtonI = loveframes.Create("imagebutton")
	backButtonI:SetSize(150, 40)
	backButtonI:SetImage("img/backButton.png")
	backButtonI:SetPos((gr.getWidth()-150)/2 , (gr.getHeight()+220)/2)
	backButtonI:SetState("help")
	backButtonI:SetText("")
	backButtonI.OnClick = function()
		loveframes.SetState("startmenu")
		gamestate = "startmenu"
	end

	--GAME OVER MENU

	local tryAgainButton = loveframes.Create("imagebutton")
	tryAgainButton:SetSize(150, 40)
	tryAgainButton:SetImage("img/tryButton.png")
	tryAgainButton:SetPos((gr.getWidth()-150)/2 , (gr.getHeight()+80)/2)
	tryAgainButton:SetState("gameover")
	tryAgainButton:SetText("")
	tryAgainButton.OnClick = function()
		p:spawnAgain()
		p.score = 0
		f:respawn()
		e:newGame()
		gamestate = "playing"
		loveframes.SetState("playing")
		menuMusic:stop()
	end

	local screenshotButton = loveframes.Create("imagebutton")
	screenshotButton:SetSize(150, 40)
	screenshotButton:SetImage("img/screenshot.png")
	screenshotButton:SetPos((gr.getWidth()-150)/2 , (gr.getHeight()+200)/2)
	screenshotButton:SetState("gameover")
	screenshotButton:SetText("")
	screenshotButton.OnClick = function()
		screenshot = gr.newScreenshot(false)
		screenshot:encode("score.png","png")
		local frame = loveframes.Create("frame")
		frame:SetName("Screenshot saved")
		frame:SetSize(410, 80)
		frame:SetState("gameover")
		frame:SetPos((gr.getWidth()-410)/2, (gr.getHeight()-130)/2)

		saveSS()

		local text = loveframes.Create("text", frame)
		text:SetText("Screenshot saved as \"score.png\" in the game directory")
		text:SetPos(10, 30)
	end

	local backButtonG = loveframes.Create("imagebutton")
	backButtonG:SetSize(150, 40)
	backButtonG:SetImage("img/backButton.png")
	backButtonG:SetPos((gr.getWidth()-150)/2 , (gr.getHeight()+320)/2)
	backButtonG:SetState("gameover")
	backButtonG:SetText("")
	backButtonG.OnClick = function()
		p:spawnAgain()
		en:newGame()
		gamestate = "startmenu"
		loveframes.SetState("startmenu")
		if mute == false then menuMusic:play() end
	end
end